Arm-64: Add initial support for PAC-RET

SwapnilGaikwad opened the pull request:
This PR introduces initial support for Pointer Authentication (PAC) on Arm64. PAC is a hardware security feature designed to mitigate Return-Oriented Programming (ROP) attacks by cryptographically signing return addresses. The signed return address is stored on the stack and later authenticated before returning from a function, ensuring control flow returns to the intended caller.

More details on PAC and its role in software security can be found ([here](https://llsoftsec.github.io/llsoftsecbook/#pointer-authentication)).

Enabling PAC involves inserting additional instructions into both the function prolog (for signing) and epilog (for authentication). This results in increased code size. For example, we observe a 1.8% increase in code size across System*.dll assemblies compiled using crossgen2.

The added instructions also introduce some runtime overhead. In our benchmark of Orchard CMS 9.0, we observe a 1.3% performance regression, which falls within the noise range (standard deviation: ~1.3%).

@kunalspathak @janvorli @a74nh 

Contributes to https://github.com/dotnet/runtime/issues/109457

SwapnilGaikwad committed:
Add auth-strip in prolog and epilog of jitted methods

kunalspathak commented:
@SwapnilGaikwad - once you resolve the stack walking problem, can you also try something like this in a Foo() method?

```asm
paciasp
stp     fp, lr, [sp, #-0x10]!

mov x9, [sp] ;  overwrite lr value with random content
...
...
xpaclri  ; will fail and should give the call stack
...
```

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Use unwind info to indicate use of PAC

- Add support for 'pac_sign_lr' unwind code
- Use authenticate instruciton instead of stripping PAC in epilog

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Switch to auth with zero and fix tail call optimisations

kunalspathak committed:
fix nativeaot CFI unwind code

kunalspathak committed:
fix build

janvorli commented on src/coreclr/jit/unwindarm64.cpp:
```diff
@@ -638,14 +638,26 @@ void Compiler::unwindSaveNext()
+        }
+
+        // DW_CFA_GNU_window_save 0x2D
+        createCfiCode(func, cbProlog, CFI_DEF_CFA_REGISTER, DWARF_REG_ILLEGAL);
```
I am not sure how the arguments CFI_DEF_CFA_REGISTER, DWARF_REG_ILLEGAL translate to the DW_CFA_GNU_window_save.

kunalspathak replied:
It doesn't. This is a temporary thing to unblock other CI jobs. I am trying to figure out how to come up with the encode to pass to `createCfiCode()` . I do not see codes that are being passed to that method.

kunalspathak committed:
Add STRIP_PAC

kunalspathak committed:
wip

kunalspathak committed:
Uncomment the implementation of RtlStripPacOnline()

kunalspathak committed:
Add some missing declaration of CFI_NEGATE_RA_STATE

kunalspathak committed:
build fix

kunalspathak committed:
Comment StripOnline

kunalspathak committed:
Add JitPacEnabled() switch

kunalspathak committed:
Make PacEnabled a Release flag

kunalspathak committed:
Just strip from bit 48 onwards because thats what works for linux

SwapnilGaikwad committed:
Fix return address hijacking

kunalspathak commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -4545,6 +4545,30 @@ struct ExecutionState
+#if defined(TARGET_ARM64)
+static inline void* PacSignPtr(void* ptr)
+{
+    __asm__ volatile (".arch_extension pauth\n"
```
i guess to compile this, you need to pass the flags that you intended to pass in https://github.com/dotnet/runtime/pull/108561, right?

SwapnilGaikwad replied:
On Linux, it shouldn't require compile time flag as we are specifying that path is required in assembly itself. This would need to be refactored when we will merge #108561. However, I'm not sure yet about the changes to inline assembly that we need to make builds happy on Windows.

kunalspathak committed:
Unwinder: Strip the Lr before storing in PC

SwapnilGaikwad committed:
Emit pac unwind info in epilog

kunalspathak commented:
/azp run runtime-coreclr jitstress, gcstress0x3-gcstress0xc

kunalspathak commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

SwapnilGaikwad committed:
Add inline assembly variants for MSVC

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Remove MSVC assembly

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Restore previous fix to check its impact on nativeaot failures

SwapnilGaikwad committed:
Fix basic AOT tests

SwapnilGaikwad committed:
Fix reaturn address hijacking for AOT

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Fix return address comparison for tail calls in native aot

SwapnilGaikwad committed:
Sign return address while restoring it in unhijacking a thread

SwapnilGaikwad committed:
Disable emitting authenticate to see behaviour on windows

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Fix build errors

SwapnilGaikwad committed:
Fix build errors

SwapnilGaikwad committed:
Fix build errors for casting

SwapnilGaikwad committed:
Fix build errors

SwapnilGaikwad committed:
Merge main

kunalspathak commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc, runtime-coreclr jitstress

kunalspathak commented:
The osx-arm64 failure is coming from https://github.com/dotnet/runtime/pull/110472/files#diff-2d7e5e7d3e0060ca5872dcefd6e716c18227def926ed32375c01cdf5a698c1aeR655. Need to fix it properly.

kunalspathak commented:
/azp run runtime-coreclr gcstress-extra, runtime-coreclr libraries-jitstress, runtime-coreclr outerloop

kunalspathak commented:
- outerloop : no PAC related failures
- gcstress-extra: no PAC related failures
- runtime: osx/arm64 failures related to CFI as pointed in https://github.com/dotnet/runtime/pull/110472#issuecomment-2776160360
- libraries-jitstress: Some failures related to tail call [logs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1005615&view=ms.vss-test-web.build-test-results-tab&runId=26839806&paneView=debug).

SwapnilGaikwad committed:
Put dwarf cfi code for pac

SwapnilGaikwad committed:
Add a placeholder for OSX ObjectWriter

kunalspathak commented:
/azp run runtime-coreclr crossgen2, runtime-coreclr crossgen2 outerloop

filipnavara commented:
> We think it might be related to the code ObejectWriter, potentially `MachObjectWriter.cs`.

MachObjectWriter has a logic to convert some DWARF sequences into a single compact opcode. It should either work or bail out and fallback to the same logic that Linux uses.

I suppose the best way to figure out what is happening is to download the Helix artifacts with the `runfo` tool (`runfo get-helix-payload -j 514779b9-ff18-49e0-a531-ca2b6e761146 -w System.Runtime.Tests -o helix`) and inspect them with `llvm-dwarfdump`.

SwapnilGaikwad commented:
> I suppose the best way to figure out what is happening is to download the Helix artifacts with the `runfo` tool (`runfo get-helix-payload -j 514779b9-ff18-49e0-a531-ca2b6e761146 -w System.Runtime.Tests -o helix`) and inspect them with `llvm-dwarfdump`.

I tried the `dwarfdump` (Apple LLVM version 17.0.0) but it fails to dump with the following error.
```
.eh_frame contents:
error: parsing entry instructions at 0x18 failed
```

filipnavara commented:
> Would you recommend any other tool or a config variable to produce more info to aid debugging?

I'll have a look today or tomorrow.

filipnavara commented:
One of the differences between the ILC handling of the DWARF data is that on Linux it passes through the `CompressARM64CFI` method for preprocessing. That essentially builds the whole layout of a frame and then rewrites the DWARF info from scratch to represent the frame at the code offset equal to the end of the prolog. Since we don't do unwinding in prologs it produces more compact representation that carries around the necessary information.

On Apple platforms we don't do this compression step. Back when Apple support was added it caused problems with the Apple linkers, the LLVM object writer (which has since been replaced), and the logic in those that tried to produce the compact unwinding information.

I can see that the emitted info produces a `DW_CFA_def_cfa` opcode with `0xff` value for the register which is indeed invalid. Presumably this would only happen if the JIT produces the `CFI_DEF_CFA` code which it never does, which gets us to the real problem...

... the definition of `CFI_NEGATE_RA_STATE` in `CFI_OPCODE` in JIT doesn't match the one in the managed code because the one in the JIT doesn't have the `CFI_DEF_CFA` value in the enum.

filipnavara commented on src/coreclr/inc/cfi.h:
```diff
@@ -9,7 +9,8 @@ enum CFI_OPCODE
    CFI_ADJUST_CFA_OFFSET,    // Offset is adjusted relative to the current one.
    CFI_DEF_CFA_REGISTER,     // New register is used to compute CFA
-   CFI_REL_OFFSET            // Register is saved at offset from the current CFA
+   CFI_REL_OFFSET,           // Register is saved at offset from the current CFA
+   CFI_NEGATE_RA_STATE,      // Sign the return address in lr with pacibsp
```
```suggestion
   CFI_REL_OFFSET,           // Register is saved at offset from the current CFA
   CFI_DEF_CFA,              // Used by ILC, not emitted by JIT
   CFI_NEGATE_RA_STATE,      // Sign the return address in lr with pacibsp
```

jkotas commented on src/coreclr/jit/emitarm64.cpp:
```diff
@@ -1397,6 +1397,34 @@ static const char * const  bRegNames[] =
+    {
+        return;
+    }
+    // TODO-PAC: should be paciasp
```
I see that you have `Sign with SP` at the end of the list. I am not sure whether you want to wait with tackling it as the last item. I expect that you will need to revisit and retest number of changes in this PR to make signing with SP work.

kunalspathak replied:
I agree. I don't think we will do signing with `SP` just yet and will be in future PR once we are confident about the basic PAC working we implemented.

jkotas replied:
Signing return addresses with `SP` is required to deliver on security promise of PAC and it changes how things like hijacking need to be handled. I expect that you will need to redo a large part of change to make it work and then retest everything again. You can certainly do that, but I do not think it is the most efficient way to deliver this feature.

a74nh replied:
First, I'm assuming https://github.com/dotnet/runtime/pull/108561 will be merged into this to give PAC across C++ and C# code.

> Signing return addresses with `SP` is required to deliver on security promise of PAC and it changes how things like hijacking need to be handled.

This depends what we're protecting against. In a full ROP attack an attacker wants to write arbitrary addresses to the stack. They want to use small gadgets, typically one or two instructions followed by a ret, that just achieves a single action (eg add + ret). Attackers build up a library of gadgets, which can then be chained together on the stack, and essentially run their own program.

By using signing with zero the attacker cannot write arbitrary addresses to the stack (because they would need to be signed and the attacker has no way to sign arbitrary addresses).

However, signing with 0 allows addresses that have already been signed to be used elsewhere. So, the attacker could regularly scan the stack and build up a collection of signed addresses. These will point to the return points of any functions which have previously been called in the running program. There is definitely the potential for there to be useful gadgets here, for example when the return from a function is then very close to a return. Eg the return from foo here gives a gadget to increment r0 assuming the arbitrary code doesn't do much.
```csharp
 .....
 foo(...);
 ..... arbitrary code ....
 return ret+1;
 ....
```
But most return locations won't be of practical use. It's much less likely for there to be enough gadgets to string together enough functionality to do what the hacker wants. Also, the attacker needs to rebuild this list of addresses every time the program is run (as the key changes per process).

Signing with SP will remove the option to use these return point gadgets.

The C++ code will be signed with SP (via #108561). Which means with this PR the attacker is restricted to return point gadgets in C# code only.

Having written this out, I've convinced myself there still is a potential for attack when using 0, but, it's very impractical.

SwapnilGaikwad committed:
Add DEF_CFA opcode for ilc

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad commented:
> I can see that the emitted info produces a `DW_CFA_def_cfa` opcode with `0xff` value for the register which is indeed invalid. Presumably this would only happen if the JIT produces the `CFI_DEF_CFA` code which it never does, which gets us to the real problem...

Interesting! Thanks a lot for spotting and providing a fix.

SwapnilGaikwad committed:
Fix formatting

a74nh commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc, runtime-coreclr jitstress

SwapnilGaikwad committed:
Merge main

AndyAyersMS commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc, runtime-coreclr jitstress

SwapnilGaikwad commented:
From the stress test results the jitstress failure is on x64 so seems unrelated. The GCsterss failure on OSX+arm64 seems to be as a result of timeout than a failure.
Now I'll merge #108561 to enable PAC for C/C++ code.

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Enable PAC for non-jitted code

kunalspathak commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc, runtime-coreclr jitstress, runtime-coreclr libraries-jitstress2-jitstressregs, runtime-coreclr outerloop

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Intial clean-up: use xpacia instead of a mask

kunalspathak commented on src/coreclr/nativeaot/Runtime/StackFrameIterator.cpp:
```diff
@@ -1599,6 +1599,19 @@ void StackFrameIterator::UnwindUniversalTransitionThunk()
+{
+    __asm__ volatile (".arch_extension pauth\n"
+                      "xpaci %0"
+                      : "+r" (ptr)
```
can we move this in `asmhelpers.S` and `asmhelpers.asm`?

SwapnilGaikwad replied:
Done.

kunalspathak commented on eng/native/configurecompiler.cmake:
```diff
@@ -705,6 +705,11 @@ if (CLR_CMAKE_HOST_UNIX)
 
+  if(CLR_CMAKE_HOST_ARCH_ARM64 AND CLR_CMAKE_HOST_LINUX)
+    # Enable Pointer Authentication (PAC) extension
+    add_compile_options(-mbranch-protection=pac-ret)
```
can we just pass this as part of `EXTRA_CFLAGS` or `EXTRA_CXXFLAGS` and document it? that way PAC will stay off by default.

SwapnilGaikwad replied:
Done

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Clean-up inline assembly

SwapnilGaikwad committed:
Remove PAC for non-jitted code, added using compile time flags

SwapnilGaikwad committed:
Fix windows build errors

SwapnilGaikwad committed:
Fix build errors

SwapnilGaikwad commented on src/coreclr/unwinder/arm64/unwinder.cpp:
```diff
@@ -33,12 +33,13 @@
-#ifdef HOST_UNIX
-#define RtlZeroMemory ZeroMemory
-
 #if defined(TARGET_ARM64)
```
@kunalspathak should this be `if !defined(DACCESS_COMPILE) && defined(TARGET_ARM64)` ? The same on [Line:273](https://github.com/dotnet/runtime/pull/110472/commits/a99f6391e0a6ea4ce6036e8aae87a20b18d2c195#diff-8d8eea22f227bfd092bc709d426a8e62e6d68ec03c03fcd5b36b579bb78ba729R273)

kunalspathak replied:
The only place where `PacStripPtr` is getting used in this file is in ` !defined(DACCESS_COMPILE)`, so doesn't harm to keep it that way...so you might need `#if defined(TARGET_ARM64) && !defined(DACCESS_COMPILE)` here.

kunalspathak commented on src/coreclr/vm/arm64/asmhelpers.S:
```diff
@@ -337,6 +337,13 @@ NESTED_END OnHijackTripThread, _TEXT
 #endif // FEATURE_HIJACK
 
+// void* PacStripPtr(void *);
+.arch_extension pauth
```
do you have machine that doesn't have PAC, where you can validate if this will not break?

SwapnilGaikwad replied:
yup, I'll give it go.

SwapnilGaikwad replied:
This was a good shout as `xpaci` doesn't fall in the nop/hint space. This leads to SIGILL on non-PAC machines. Now switched to using `xpaclri` that is nop on non-PAC machines. Confirmed this by running all tests on a non-PAC machine.

kunalspathak replied:
From https://devblogs.microsoft.com/oldnewthing/20220819-00/?p=107020, I see that it points that to use `xpaclri`, we should check for feature support. I am wondering if it is possible that we get `SIGILL` on different machines where we didn't test this out like surface, etc. 

![image](https://github.com/user-attachments/assets/93c1ee18-c792-4dee-ac10-cb55bc103747)

SwapnilGaikwad replied:
Sure, we can test it on another non-PAC machine. 
The `xpacd` and `xpaci` instructions don't fall into the nop/hint space but `xpaclri` does [1], so using xpaclri should be fine on non-pac machines..
![image](https://github.com/user-attachments/assets/21c5f4ec-e22f-4d49-848f-4cd87b18cade)
![image](https://github.com/user-attachments/assets/3efb98ae-1796-473c-9367-13d44d645d5c)

[1] https://docsmirror.github.io/A64/2024-03/xpac.html

kunalspathak commented on src/coreclr/vm/arm64/asmhelpers.asm:
```diff
@@ -676,6 +676,13 @@ COMToCLRDispatchHelper_RegSetup
+; void* PacStripPtr(void *);
+    LEAF_ENTRY PacStripPtr
+        ; xpaci    x0
+        DCD     0xDAC143E0
```
interestingly, for windows, you don't have to check the availability of feature like we do on linux with `.arch_extension pauth`?

SwapnilGaikwad replied:
The .arch_extension is primary to make clang emit the PAC instruction in binary. Here we directly put the binary equivalent of `xpaci x0` so compiler that doesn't support PAC yet, it won't complain about it. It doesn't cause harm on non-pac systems as the PAC instructions would fall into the nop space.

kunalspathak commented on src/coreclr/vm/arm64/asmhelpers.asm:
```diff
@@ -676,6 +676,13 @@ COMToCLRDispatchHelper_RegSetup
+; void* PacStripPtr(void *);
+    LEAF_ENTRY PacStripPtr
+        ; xpaci    x0
+        DCD     0xDAC143E0
```
`0xDAC143E0` <- Is that opcode? can you write a comment about it?

SwapnilGaikwad replied:
This is the binary version of the `xpaci    x0` instruction. I'll elaborate the comment above.

SwapnilGaikwad replied:
Put some comments about using the binary version as well using the `xpaclri` instead of `xpaci`.

kunalspathak commented on src/coreclr/jit/jitconfigvalues.h:
```diff
@@ -125,6 +125,7 @@ CONFIG_STRING(JitInlineMethodsWithEHRange, "JitInlineMethodsWithEHRange")
 
 CONFIG_INTEGER(JitLongAddress, "JitLongAddress", 0) // Force using the large pseudo instruction form for long address
 CONFIG_INTEGER(JitMaxUncheckedOffset, "JitMaxUncheckedOffset", 8)
+RELEASE_CONFIG_INTEGER(JitPacEnabled, "JitPacEnabled", 1)
```
a remainder to turn it off before merging.

SwapnilGaikwad replied:
Done

kunalspathak commented on src/coreclr/jit/emitarm64.cpp:
```diff
@@ -1397,6 +1397,34 @@ static const char * const  bRegNames[] =
+    {
+        return;
+    }
+    // TODO-PAC: should be autiasp
```
just to confirm, using `paciasp` and `autiasp` will happen in future PR(s), right?

SwapnilGaikwad replied:
Yes. I'll remove these todos. We can search for `paciaz` to find these places.

kunalspathak commented on src/coreclr/nativeaot/Runtime/thread.cpp:
```diff
@@ -786,6 +790,20 @@ void Thread::HijackReturnAddress(NATIVE_CONTEXT* pSuspendCtx, HijackFunc* pfnHij
 }
 
+#if defined(TARGET_ARM64) && defined(__GNUC__)
+static inline void* PacSignPtr(void* ptr)
```
We do not need this method because you have already declared this at the top?

SwapnilGaikwad replied:
Done

kunalspathak commented on src/coreclr/nativeaot/Runtime/thread.cpp:
```diff
@@ -817,6 +839,9 @@ void Thread::HijackReturnAddressWorker(StackFrameIterator* frameIterator, Hijack
 
         *ppvRetAddrLocation = (void*)pfnHijackFunction;
+#if defined(TARGET_ARM64) && defined(__GNUC__)
+        *ppvRetAddrLocation = PacSignPtr(*ppvRetAddrLocation);
```
so what happens when we do not sign the addresses? Do they go back on stack where the attacker can hijack?

kunalspathak replied:
Also, why is it just for `__GNUC__`? should also do it for windows, right? Likewise in `threadsuspend.cpp`

kunalspathak commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:
```diff
@@ -511,7 +511,7 @@ private static unsafe void DispatchTailCalls(
             TailCallTls* tls = GetTailCallInfo(callersRetAddrSlot, &callersRetAddr);
             PortableTailCallFrame* prevFrame = tls->Frame;
-            if (callersRetAddr == prevFrame->TailCallAwareReturnAddress)
+            if ((callersRetAddr) == (prevFrame->TailCallAwareReturnAddress))
```
please revert these changes.

SwapnilGaikwad replied:
Done

kunalspathak commented:
can you somewhere document the flags that we should pass in windows/linux to enable PAC for runtime? May be in [building workflow](https://github.com/dotnet/runtime/blob/main/docs/workflow/building/coreclr/README.md) document.

SwapnilGaikwad committed:
Update strip/sign mechanism to use nop variants of PAC instructions

SwapnilGaikwad committed:
Merge main

kunalspathak commented:
Few more suggestions:
- Looking at https://github.com/dotnet/runtime/pull/110472/commits/eefb18be3996f6a2041deaace859b6992d4994b7, the fact that we have to use the NOP compliant `xpaclri` version to strip/sign, we will be executing more instructions even on PAC machine that can easily just do `xpaci    x0`. So, can we have two versions of strip/sign methods? The optimized version (`xpaci x0`) will be used for JIT scenarios on PAC machine or NativeAOT/crossgen2, if target specification is `--enable-pac`. The unoptimized version (`xpaclri`) will be called for everything else.

- We will need to add a flag for nativeaot/crossgen2 `--target-pac-machine` or equivalent to give control to user if they want PAC or not.
- We should also check that for crossgen2 scenario, if user opt out of PAC, but is running on PAC where coreclr and JIT code is with PAC, how do they interact with each other.
- Currently, we just have an environment variable to enable/disable PAC feature. We also need a feature detection flag from hardware if PAC is supported or not. Then we will use something like `isPacAvailableOnHardware() && isEnvVarPacEnabled()` then enable PAC. A separate PR for it will be good to have.
- We should verify the basic debugger scenarios by live/offline debugging and note down the behavior of sos and what scenarios are broken. It will help us to evaluate if we can address them during .NET 10 and see if there is a potential to make PAC on by default.
- Can you share the disk footprint impact by PAC? You can collect that information by running crossgen2 on release version of `System*.dll` with and without PAC.
- Can you run some TE/Orchard benchmarks to see how much is the perf impact from PAC?

jkotas commented:
> We should verify the basic debugger scenarios by live/offline debugging and note down the behavior of sos and what scenarios are broken. 

You will need to redo this once you enable signing with SP. You may want to enable the signing with SP sooner rather than later so that you do not need to redo this validation.

(Also, signing with SP needs to be enabled before this is turned on by default.)

jkotas commented on src/coreclr/unwinder/arm64/unwinder.cpp:
```diff
@@ -265,16 +269,71 @@ do {
-//
 // Macros for stripping pointer authentication (PAC) bits.
-//
+#if !defined(DACCESS_COMPILE) && defined(TARGET_ARM64)
```
```suggestion
#if !defined(DACCESS_COMPILE) && !defined(FEATURE_CDAC_UNWINDER)`
```
We have two builds of out-of-proc unwinders: DACCESS and CDAC. I think we want to take the offline path for both out-of-proc unwinders. `TARGET_ARM64` should not be needed in the condition once you do enable to the offline path for all out-of-proc unwinders.

(All similar ifdefs in this file should be changed like this.)

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Incorporate review comments

SwapnilGaikwad commented:
> * Can you share the disk footprint impact by PAC? You can collect that information by running crossgen2 on release version of `System*.dll` with and without PAC.

Size of System*dll using coressgen2:
```
main -> 48156k
pac  -> 49036k   (1.8% increase)
```

> * Can you run some TE/Orchard benchmarks to see how much is the perf impact from PAC?

Orchard CMS 9.0:
```
      Average (Requests/sec)	StdDev	Overhead
main  5163.42               	1.27%	0.00%
pac   5096.84               	1.29%	1.29%
```

kunalspathak commented:
Thanks for sharing the numbers @SwapnilGaikwad .

> Orchard CMS 9.0:

out of curiosity, what do they look like on non-pac machines? do they similar drop of 1.2%? I am assuming not.

kunalspathak commented on src/coreclr/jit/codegenarmarch.cpp:
```diff
@@ -4731,6 +4731,9 @@ void CodeGen::genPushCalleeSavedRegisters()
 #endif // DEBUG
 
+    // Sign LR as part of Pointer Authentication (PAC) support
+    GetEmitter()->emitPacInProlog();
```
Either in this PR or in a follow-up PR, you might want to consider moving the pac instructions closer to where we save `lr` on the stack instead of doing it at the top all the time. That way, you could also have a way to get the `sp` address where the `lr` was saved and then use that address to build up the original `sp` used to strip/auth the `lr` in runtime layer.

![image](https://github.com/user-attachments/assets/e11e3861-5ced-4f16-98a2-62a4057669a2)

![image](https://github.com/user-attachments/assets/820bc463-eeac-4ddd-851c-aceb0d6cd323)

jkotas replied:
Yes, number of things will need to be redone to get SP signing in place.

kunalspathak commented on src/coreclr/vm/arm64/asmhelpers.S:
```diff
@@ -343,6 +343,32 @@ NESTED_END OnHijackTripThread, _TEXT
 
 #endif // FEATURE_HIJACK
 
+// void* PacStripPtr(void *);
```
Something to think about for future PR - should we always authenticate instead of stripping the return address?

kunalspathak commented:
/azp run runtime-coreclr jitstress, runtime-coreclr gcstress0x3-gcstress0xc, runtime-coreclr gcstress-extra, runtime-coreclr outerloop

kunalspathak commented:
@jkotas - any other pipeline that you recommend running once again?

jkotas commented on src/coreclr/nativeaot/Runtime/thread.cpp:
```diff
@@ -817,6 +826,9 @@ void Thread::HijackReturnAddressWorker(StackFrameIterator* frameIterator, Hijack
 #endif
 
         *ppvRetAddrLocation = (void*)pfnHijackFunction;
+#if defined(TARGET_ARM64)
+        *ppvRetAddrLocation = PacSignPtr(*ppvRetAddrLocation);
+#endif // TARGET_ARM64
```
```suggestion
        void* pvHIjackAddr = (void*)pfnHijackFunction;
#if defined(TARGET_ARM64)
        pvHIjackAddr = PacSignPtr(pvHIjackAddr);
#endif // TARGET_ARM64
        *ppvRetAddrLocation = pvHIjackAddr;
```
We should avoid writing the wrong value first, and then overwriting it with the correct value. It can cause interesting race conditions.

SwapnilGaikwad replied:
Done

jkotas commented on src/coreclr/vm/tailcallhelp.cpp:
```diff
@@ -26,6 +29,13 @@ FCIMPL2(void*, TailCallHelp::GetTailCallInfo, void** retAddrSlot, void** retAddr
+#if defined(TARGET_ARM64)
+    *retAddr = PacStripPtr(*retAddr);
+    void* tailCallAwareReturnAddress = thread->GetTailCallTls()->GetFrame()->TailCallAwareReturnAddress;
+    tailCallAwareReturnAddress = PacStripPtr(tailCallAwareReturnAddress);
```
`tailCallAwareReturnAddress` is unused. Why are we doing this?

jkotas commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -4600,6 +4605,9 @@ void Thread::HijackThread(ExecutionState *esb X86_ARG(ReturnKind returnKind) X86
     // Remember the place that the return would have gone
     m_pvHJRetAddr = *esb->m_ppvRetAddrPtr;
+#if defined(TARGET_ARM64)
+    m_pvHJRetAddr = PacStripPtr(m_pvHJRetAddr);
```
Same as my other comment: Avoid storing the wrong value and overwriting it with the correct value.

jkotas commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -4641,6 +4653,9 @@ void Thread::UnhijackThread()
         // restore the return address and clear the flag
         *m_ppvHJRetAddrPtr = m_pvHJRetAddr;
+#if defined(TARGET_ARM64)
+        *m_ppvHJRetAddrPtr = PacSignPtr(*m_ppvHJRetAddrPtr);
```
Same here

SwapnilGaikwad replied:
Done

jkotas commented on src/coreclr/vm/threadsuspend.cpp:
```diff
@@ -4600,6 +4605,9 @@ void Thread::HijackThread(ExecutionState *esb X86_ARG(ReturnKind returnKind) X86
     // Remember the place that the return would have gone
     m_pvHJRetAddr = *esb->m_ppvRetAddrPtr;
+#if defined(TARGET_ARM64)
+    m_pvHJRetAddr = PacStripPtr(m_pvHJRetAddr);
```
Would it make sense to keep the bits on `m_pvHJRetAddr` unstripped so that the original return address is protected at all times? I think stripping the bits here creates a hole in the protection. Should we strip the bits later when we read `m_pvHJRetAddr` for stackwalking, but not when we use it to store/restore the original return address so that the return address is protected for the whole time?

SwapnilGaikwad replied:
It's a good idea. Ideally, we should authenticate instead of stripping to ensure that we are returning to a valid return address. I'll check how can we do this.

jkotas replied:
> the original address is preserved so that after GC we can return to it, and the updated address would take the execution flow to desired new address

Right. My point is that it would be better to keep the signed address intact through the whole process:
- Delete PacStripPtr in Thread::HijackThread and store the original signed address in m_pvHJRetAddr instead 
- Delete PacSignPtr in Thread::UnhijackThread
- Add PacStripPtr as necessary to places that read m_pvHJRetAddr and do not expect the signature in the upper bits

It will make the original return address protected while it is stored in m_pvHJRetAddr. As implemented currently, the return address is not protected while it is stored in m_pvHJRetAddr.

jkotas commented:
> @jkotas - any other pipeline that you recommend running once again?

If this is checked in as disabled, I do not have concerns about in-depth testing.

I left some codereview comments.

SwapnilGaikwad commented on src/coreclr/unwinder/arm64/unwinder.cpp:
```diff
@@ -265,16 +269,71 @@ do {
+
+--*/
+{
+    *Pointer &= 0x0000FFFFFFFFFFFF;
```
Yes, this is the stripping mechanism that should be the same when SP is used as salt.

SwapnilGaikwad commented:
> Thanks for sharing the numbers @SwapnilGaikwad .
> 
> > Orchard CMS 9.0:
> 
> out of curiosity, what do they look like on non-pac machines? do they similar drop of 1.2%? I am assuming not.

On a non-PAC machine (altra), overhead is not noticable.
```
Average	    StdDev	Overhead
3001.04     0.87%	0.00%
3000.90     0.70%	0.00%
```

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Incorporate review comments

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Disable JitPacEnabled flag by default

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Use unwind info to detect if PAC is present for jitted code

SwapnilGaikwad committed:
Fix assert failure for return address

SwapnilGaikwad committed:
Make IsPacPresent() Arm64 only

SwapnilGaikwad committed:
Fix build errors

SwapnilGaikwad committed:
Fix build errors

SwapnilGaikwad committed:
Add a prototype of using unwind info to detect PAC

SwapnilGaikwad committed:
Merge main

jkotas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/Dwarf/DwarfCfiOpcode.cs:
```diff
@@ -11,6 +11,7 @@ internal enum CFI_OPCODE
         CFI_REL_OFFSET,           // Register is saved at offset from the current CFA
-        CFI_DEF_CFA               // Take address from register and add offset to it.
+        CFI_DEF_CFA,              // Take address from register and add offset to it.
+        CFI_NEGATE_RA_STATE,      // Sign the return address in lr with pacibsp
```
Should it be `paciasp`? I believe that it is the default used for signing return addresses.

SwapnilGaikwad replied:
'll change it to paciasp/pacibsp after adding signing with sp feature.

jkotas commented on src/coreclr/inc/cfi.h:
```diff
@@ -9,7 +9,9 @@ enum CFI_OPCODE
    CFI_DEF_CFA_REGISTER,     // New register is used to compute CFA
-   CFI_REL_OFFSET            // Register is saved at offset from the current CFA
+   CFI_REL_OFFSET,           // Register is saved at offset from the current CFA
+   CFI_DEF_CFA,              // Used by ILC, not emitted by JIT
```
```suggestion
   CFI_DEF_CFA,              // Take address from register and add offset to it.
```
All CFI opcodes are only used by AOT compilation

jkotas commented on src/coreclr/nativeaot/Runtime/windows/CoffNativeCodeManager.cpp:
```diff
@@ -831,6 +831,60 @@ bool CoffNativeCodeManager::IsUnwindable(PTR_VOID pvAddress)
+bool CoffNativeCodeManager::IsPacPresent(MethodInfo *    pMethodInfo,
+                                         REGDISPLAY *    pRegisterSet)
+{
+#if defined(HOST_WINDOWS)
```
This whole file is only used on Windows, so this ifdef is always true

jkotas commented on src/coreclr/unwinder/arm64/unwinder.cpp:
```diff
@@ -1689,72 +1751,35 @@ Routine Description:
 --*/
-
 {
-    ULONG AccumulatedSaveNexts;
```
This file contains the original implementation of the unwinders for Windows unwind info. We have been trying to keep it in sync with the master copy so that it is easy to pick up bug fixes, but this refactoring seems to diverging it quite a bit. Is this refactoring necessary to achieve what you need?

cc @dotnet/dotnet-diag

SwapnilGaikwad replied:
Cool, I'll give a go 👍

SwapnilGaikwad replied:
Done. Thanks, Jan.

jkotas commented on src/coreclr/inc/cfi.h:
```diff
@@ -9,7 +9,9 @@ enum CFI_OPCODE
-   CFI_REL_OFFSET            // Register is saved at offset from the current CFA
+   CFI_REL_OFFSET,           // Register is saved at offset from the current CFA
+   CFI_DEF_CFA,              // Used by ILC, not emitted by JIT
+   CFI_NEGATE_RA_STATE,      // Sign the return address in lr with paciaz
```
Does the spec for `DW_CFA_AARCH64_negate_ra_state` opcode require specific signing or is it meant to support any type of signing?

In other words, are there any unwinders that will break when we use non-standard signing with this opcode?

SwapnilGaikwad replied:
That's a good point. It's likely that an unwinder could break. I'll investigate this further. Preliminary looks suggests that [gcc's unwinder](https://github.com/gcc-mirror/gcc/blob/master/libgcc/config/aarch64/aarch64-unwind.h#L37-L41) reconises only unsinged or signing with SP modes.

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Extract IsPacPresent() outside the undwinder.cpp

jkotas commented on src/coreclr/inc/clrnt.h:
```diff
@@ -369,6 +369,20 @@ RtlVirtualUnwind(
     IN OUT PKNONVOLATILE_CONTEXT_POINTERS ContextPointers OPTIONAL
     );
 
+EXTERN_C
```
Revert

SwapnilGaikwad replied:
Done

SwapnilGaikwad committed:
Remove clrnt.h changes

SwapnilGaikwad committed:
Address review comments

SwapnilGaikwad commented on src/coreclr/vm/excep.cpp:
```diff
@@ -6930,6 +6930,80 @@ bool IsIPInEpilog(PTR_CONTEXT pContextToCheck, EECodeInfo *pCodeInfo, BOOL *pSaf
+    // We should always get a function entry for a managed method
+    _ASSERTE(FunctionEntry != NULL);
+    DWORD_PTR ImageBase = pCodeInfo->GetModuleBase();
+    ULONG_PTR UnwindDataPtr = (ULONG_PTR)(ImageBase + FunctionEntry->UnwindData);
```
Done.

SwapnilGaikwad commented:
Hi @jkotas , I was wondering if you any suggestions to investigate a segfault. The NativeAot version of `tests/GC/API/GC/GetTotalAllocatedBytes.cs` test on Linux+arm64 is hitting a segfault after jumping to an address with signed pointer. The signed address points to `RhpGcProbeHijack` function in `GcProbe.S`. This function is used as default hijack target during return address hijacking so expecting it to have unsigned address. 
The stacktrace (listed below) after the segfault suggests that an exception may have triggered unwinding. However, while debugging, it doesn't hit the `RhpCallFilterFunclet`. Printing more info for debugging causes test to pass :). 

```
frame #0: 0x0000aaaaaad46a4c GetTotalAllocatedBytes`RhpGcProbeHijack at GcProbe.S:123
frame #1: 0x0000aaaaaad46a4c GetTotalAllocatedBytes`RhpCallFilterFunclet at ExceptionHandling.S:662
```

Any pointers to debug this? I'm using a debug build. Currently checking if any GC workflow jumps to this address.

jkotas commented on src/coreclr/nativeaot/Runtime/arm64/ExceptionHandling.S:
```diff
@@ -301,6 +301,7 @@ LOCAL_LABEL(TailCallWasHijacked):
         mov         lr, x1
         str         lr, [sp, #(rsp_offsetof_Context + OFFSETOF__PAL_LIMITED_CONTEXT__LR)]
         str         lr, [sp, #(rsp_offsetof_Context + OFFSETOF__PAL_LIMITED_CONTEXT__IP)]
+        xpaclri
```
Are the tests able to hit this path?

It does not look right that we are clearing the signature for `lr` on `TailCallWasHijacked` path, but not on the regular path.

SwapnilGaikwad replied:
I'll take a look again.

jkotas commented:
> Any pointers to debug this? I'm using a debug build.

I do not see this specific failure in the CI, but I see plenty of other failures. They are related to hijacking - we are trying to return to RhpGcProbeHijack, but the return address is not signed.

> This function is used as default hijack target during return address hijacking so expecting it to have unsigned address.

The hijack target is returned up during return address hijacking. It is expected that the hijack target is signed - it is expected to happen here https://github.com/dotnet/runtime/pull/110472/files#diff-ab37b443cf4fe3a293dfbfddb896222f93c6b3c3630a6b023bb2c27399a841bdR838

> Printing more info for debugging causes test to pass :).

Yes, crashes related to GC suspension tend to be timing dependent. You may want to create a more stressful repro to debug it - run the test routing the problem on one thread and run a `for (;;) { GC.Collect(); Thread.Sleep(1); }` in a loop on background thread.

SwapnilGaikwad commented:
> The hijack target is returned up during return address hijacking. It is expected that the hijack target is signed - it is expected to happen here

Correct. However, we should jump to the hijack target after authenticating so shouldn't have it to be signed. You're right, it could be the case where we are incorrectly detecting PAC which may lead to jumping to this address without auth. I'll probably fix the detecting PAC using DWARF info and come back to this failure (hopefully it disappears :) ). I don't think currently I'm having correct starting unwind code.

> Yes, crashes related to GC suspension tend to be timing dependent. You may want to create a more stressful repro to debug it - run the test routing the problem on one thread and run a `for (;;) { GC.Collect(); Thread.Sleep(1); }` in a loop on background thread.

Cool, I'll use this if the issue persists. Thanks for the pointers 👍

SwapnilGaikwad committed:
Fix starting offset while unwinding dwarf info

SwapnilGaikwad committed:
Fix build failures of MacOS

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

risc-vv commented:
<details><summary>c9f8bd14 is being scheduled for building and testing</summary>

GIT: `c9f8bd1489bbbb1941186bf38cc9ef2f8fda2a46`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

risc-vv commented:
<details><summary>c9f8bd14 is being scheduled for building and testing</summary>

GIT: `c9f8bd1489bbbb1941186bf38cc9ef2f8fda2a46`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Fix SIGILL in NativeAOT during unwinding

SwapnilGaikwad committed:
Fix segfaults in libraries tests

SwapnilGaikwad committed:
Restore JIT's stack unwinding that was lost while restoring unwinder.cpp

AndyAyersMS commented:
Have we tried running the diagnostic tests on this yet?

risc-vv commented:
<details><summary>34401fb0 is being scheduled for building and testing</summary>

GIT: `34401fb08d91f230ca12b3ff311cdf67fd54c858`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

SwapnilGaikwad committed:
Merge main

amanasifkhalid commented:
> Have we tried running the diagnostic tests on this yet?

@steveisok is doing this now

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

SwapnilGaikwad committed:
Merge main

JulieLeeMSFT commented:
@SwapnilGaikwad, I think this PR needs some work as discussed offline. Can we change this PR to draft until it is ready to review?
